src/Dofs/DirichletBoundaryConditions.jl

```diff
@@ -0,0 +1,237 @@
+
+```jldoctest
+julia> t = 1.0;
+
+julia> update!(dbc, t)
+```
+
+The boundary conditions can be applied to a vector:
+
+```jldoctest
+julia> u = zeros(ndofs(dh))
+
+julia> apply!(u, dbc)
+```
+
+"""
+immutable DirichletBoundaryCondition
+    f::Function
+    nodes::Set{Int}
+    field::Symbol
+    components::Vector{Int}
+    idxoffset::Int
+end
+
+
+immutable DirichletBoundaryConditions{DH <: DofHandler}
+    bcs::Vector{DirichletBoundaryCondition}
+    dofs::Vector{Int}
+    free_dofs::Vector{Int}
+    values::Vector{Float64}
+    dh::DH
+    closed::Ref{Bool}
+end
+
+function DirichletBoundaryConditions(dh::DofHandler)
+    @assert isclosed(dh)
+    DirichletBoundaryConditions(DirichletBoundaryCondition[], Int[], Int[], Float64[], dh, Ref(false))
+end
+
+function Base.show(io::IO, dbcs::DirichletBoundaryConditions)
+    println(io, "DirichletBoundaryConditions:")
+    if !isclosed(dbcs)
+        print(io, "  Not closed!")
+    else
+        println(io, "  BCs:")
+        for dbc in dbcs.bcs
+            println(io, "    ", "Field: ", dbc.field, " ", "Components: ", dbc.components)
+        end
+    end
+end
+
+isclosed(dbcs::DirichletBoundaryConditions) = dbcs.closed[]
+dirichlet_dofs(dbcs::DirichletBoundaryConditions) = dbcs.dofs
+free_dofs(dbcs::DirichletBoundaryConditions) = dbcs.free_dofs
+function close!(dbcs::DirichletBoundaryConditions)
+    fill!(dbcs.values, NaN)
+    dbcs.closed[] = true
+    fdofs = Array(setdiff(dbcs.dh.dofs_nodes, dbcs.dofs))
+    resize!(dbcs.free_dofs, length(fdofs))
+    copy!(dbcs.free_dofs, fdofs)
+    return dbcs
+end
+
+function add!(dbcs::DirichletBoundaryConditions, field::Symbol,
+                          nodes::Union{Set{Int}, Vector{Int}}, f::Function, component::Int=1)
+    add!(dbcs, field, nodes, f, [component])
+end
+
+# Adds a boundary condition
+function add!(dbcs::DirichletBoundaryConditions, field::Symbol,
+                          nodes::Union{Set{Int}, Vector{Int}}, f::Function, components::Vector{Int})
+    @assert field in dbcs.dh.field_names || error("Missing: $field")
+    for component in components
+        @assert 0 < component <= ndim(dbcs.dh, field)
+    end
+
+    if length(nodes) == 0
+        warn("Added Dirichlet BC to node set containing 9 nodes")
+    end
+
+    dofs_bc = Int[]
+    offset = dof_offset(dbcs.dh, field)
+    for node in nodes
+        for component in components
+            push!(dofs_bc, dbcs.dh.dofs_nodes[offset + component, node])
+        end
+    end
+
+    n_bcdofs = length(dofs_bc)
+
+    append!(dbcs.dofs, dofs_bc)
+    idxoffset = length(dbcs.values)
+    resize!(dbcs.values, length(dbcs.values) + n_bcdofs)
+
+    push!(dbcs.bcs, DirichletBoundaryCondition(f, Set(nodes), field, components, idxoffset))
+
+end
+
+# Updates the DBC's to the current time `time`
+function update!(dbcs::DirichletBoundaryConditions, time::Float64 = 0.0)
+    @assert dbcs.closed[]
+    bc_offset = 0
+    for dbc in dbcs.bcs
+        # Function barrier
```

Comment:
If we have to do dynamic dispatch on `dbc.f` every time we call it, things will be slow. Therefore do dynamic dispatch once and call out to a function which will specialize on every `f`.